yt/utilities/on_demand_imports.py

```diff
@@ -22,6 +22,58 @@ def __call__(self, *args, **kwargs):
         raise self.error
 
 
+class OnDemandImport:
+    """
+    Implement a safely importable drop-in replacement for a module that
+    might not be installed, and fail *only* on the first attempt to use it.
+    """
+
+    def __init__(self, name: str):
+        if name.startswith("astropy"):
+            print(name)
+        self._name = name
+        self._module = None
+        self._submodules = {}
+
+    @property
+    def module(self):
+        if self._module is None:
+            try:
+                self._module = import_module(self._name)
+            except ImportError as exc:
+                raise NotAModule(self._name).error from exc
+        return self._module
+
+    def __getattr__(self, key):
+        # the recursive nature of this method is intended to support
+        # arbitrary levels of nesting for submodules that need to be
+        # explicitly imported on their own, like scipy.spatial or astropy.io.fits
+        if key in self._submodules:
+            return self._submodules[key]
+        try:
+            return getattr(self.module, key)
+        except AttributeError:
+            self._submodules[key] = OnDemandImport(f"{self._name}.{key}")
+            return self._submodules[key]
+
+
+targets = [
+    "astropy",
+    "f90nml",
+    "h5py",
+    "libconf",
+    "netCDF4",
+    "nose",
+    "pandas",
+    "pooch",
+    "requests",
+    "scipy",
+    "yaml",
+]
+for target in targets:
+    exec(f"{target} = OnDemandImport('{target}')")
```

Comment:
This should achieve the same thing as above, but imho is clearer
```suggestion
import sys
_module = sys.modules[__name__]
for target in targets:
    setattr(_module, target, OnDemandImport("target"))
```

See https://stackoverflow.com/questions/1429814/how-to-programmatically-set-a-global-module-variable for a discussion about doing this ;)

Reply:
Thanks. Indeed I was not pleased with using `exec` so I agree your solution is better. Right now I want to go across my TODO list but I'll include your suggestion before I remove the draft status !